pkg/server/server.go

```diff
@@ -232,33 +235,68 @@ func (s *Server) Write(ctx context.Context, req *openfgapb.WriteRequest) (*openf
 }
 
 func (s *Server) Check(ctx context.Context, req *openfgapb.CheckRequest) (*openfgapb.CheckResponse, error) {
+
 	store := req.GetStoreId()
 	tk := req.GetTupleKey()
 	ctx, span := s.tracer.Start(ctx, "check", trace.WithAttributes(
-		attribute.KeyValue{Key: "store", Value: attribute.StringValue(req.GetStoreId())},
+		attribute.KeyValue{Key: "store", Value: attribute.StringValue(store)},
 		attribute.KeyValue{Key: "object", Value: attribute.StringValue(tk.GetObject())},
 		attribute.KeyValue{Key: "relation", Value: attribute.StringValue(tk.GetRelation())},
 		attribute.KeyValue{Key: "user", Value: attribute.StringValue(tk.GetUser())},
 	))
 	defer span.End()
 
+	if tk.GetUser() == "" || tk.GetRelation() == "" || tk.GetObject() == "" {
+		return nil, serverErrors.InvalidCheckInput
+	}
+
 	modelID, err := s.resolveAuthorizationModelID(ctx, store, req.GetAuthorizationModelId())
 	if err != nil {
 		return nil, err
 	}
 	span.SetAttributes(attribute.KeyValue{Key: "authorization-model-id", Value: attribute.StringValue(modelID)})
 
-	q := commands.NewCheckQuery(s.datastore, s.tracer, s.meter, s.logger, s.config.ResolveNodeLimit)
+	model, err := s.datastore.ReadAuthorizationModel(ctx, store, modelID)
+	if err != nil {
+		return nil, err
+	}
 
-	res, err := q.Execute(ctx, &openfgapb.CheckRequest{
-		StoreId:              store,
-		TupleKey:             tk,
-		ContextualTuples:     req.GetContextualTuples(),
-		AuthorizationModelId: modelID,
-		Trace:                req.GetTrace(),
+	typesys := typesystem.New(model)
+
+	if err := validation.ValidateUserObjectRelation(typesys, tk); err != nil {
+		return nil, serverErrors.ValidationError(err)
+	}
+
+	for _, ctxTuple := range req.GetContextualTuples().GetTupleKeys() {
+		if err := validation.ValidateTuple(typesys, ctxTuple); err != nil {
+			return nil, serverErrors.HandleTupleValidateError(err)
+		}
+	}
+
+	g := graph.NewLocalChecker(storage.NewContextualTupleDatastore(s.datastore), 100)
+
+	ctx = typesystem.ContextWithTypesystem(ctx, typesystem.New(model))
+	ctx = storage.ContextWithContextualTuples(ctx, req.ContextualTuples.GetTupleKeys())
```

Comment:
is there a reason why this code is here and not within `graph/check.go`? Also: can you create a task in our backlog that removes the old code, if this refactor proves to be successful in prod?

Reply:
Yes, because this implementation of Check doesn't care about request scoped context like contextual tuples, and so this code shouldn't belong in `graph/check.go`.

The storage layer now respects handling ContextualTuples if they are in the context, and so the responsibility is on the caller (e.g. `server/server.go`) to inject the ContextualTuple context into the context.

This seperates responsibilities/concerns between the Check implementation and the storage layer. The storage layer returns the tuples matching both the query AND those contextual tuples matching the query.